jmclient/jmclient/blockchaininterface.py

```diff
@@ -63,14 +56,96 @@ def import_addresses_if_needed(self, addresses, wallet_name):
         """import addresses to the underlying blockchain interface if needed
         returns True if the sync call needs to do a system exit"""
 
-    def fee_per_kb_has_been_manually_set(self, N):
-        '''if the 'block' target is higher than 1000, interpret it
-        as manually set fee/Kb.
-    '''
-        if N > 1000:
-            return True
+    @abstractmethod
+    def _get_mempool_min_fee(self) -> Optional[int]:
+        """Returns minimum mempool fee as a floor to avoid relay problems
+        or None in case of error.
+        """
+
+    @abstractmethod
+    def _estimate_fee_basic(self, conf_target: int) -> Optional[int]:
+        """Returns basic fee estimation for confirmation target in blocks.
+        Additional JoinMarket fee logic is added on top, see
+        `estimate_fee_per_kb` for details. Returns feerate in sats per vB
+        or None in case of error.
+        """
+
+    def _fee_per_kb_has_been_manually_set(self, tx_fees: int) -> bool:
+        """If the block target (tx_fees) is higher than 1000, interpret it
+        as manually set fee sats/kvB.
+        """
+        return tx_fees > 1000
+
+    def estimate_fee_per_kb(self, tx_fees: int) -> int:
+        """ The argument tx_fees may be either a number of blocks target,
+        for estimation of feerate by Core, or a number of satoshis
+        per kilo-vbyte (see `_fee_per_kb_has_been_manually_set` for
+        how this is distinguished).
+        In both cases it is prevented from falling below the current
+        minimum feerate for tx to be accepted into node's mempool.
+        In case of failure to connect, source a specific minimum fee relay
+        rate (which is used to sanity check user's chosen fee rate), or
+        failure to source a feerate estimate for targeted number of blocks,
+        a default of 10000 is returned.
+        """
+
+        # default to use if fees cannot be estimated
+        fallback_fee = 10000
+
+        tx_fees_factor = abs(jm_single().config.getfloat('POLICY', 'tx_fees_factor'))
+
+        mempoolminfee_in_sat = self._get_mempool_min_fee()
+        # in case of error
+        if mempoolminfee_in_sat is None:
+            mempoolminfee_in_sat = fallback_fee
+        mempoolminfee_in_sat_randomized = random.uniform(
+            mempoolminfee_in_sat, mempoolminfee_in_sat * float(1 + tx_fees_factor))
+
+        if self._fee_per_kb_has_been_manually_set(tx_fees):
+            N_res = random.uniform(tx_fees, tx_fees * float(1 - tx_fees_factor))
```

Comment:
Here's typo bug (found by manual testing), should be `+` instead of `-`, will update soon.

Reply:
Ah right, we randomize only upwards, so a lower value is a bug. Good catch.

This is one of the reasons we should eventually de-duplicate more out of this function.

Reply:
Agree about de-duplication and also fee estimation code currently lacks any unit tests, that could have catched this.